---
title: "IsoBayes"
author:
- name: Jordy Bollon
  affiliation: 
  - Research Center dedicated to Personalized, Preventive and Predictive Medicine (CMP3VDA), Computational Department, Aosta, Italy
  email: jordy.bollon@iit.it
- name: Simone Tiberi
  affiliation:
  - Department of Statistical Sciences, University of Bologna, Bologna, Italy
  email: simone.tiberi@unibo.it
package: "`r BiocStyle::pkg_ver('IsoBayes')`"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
bibliography: References.bib
vignette: >
  %\VignetteIndexEntry{IsoBayes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
header-includes: 
- \usepackage{csquotes}
output: 
  BiocStyle::html_document
---

---
```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png",
                      message=TRUE, error=FALSE, warning=TRUE)
```

# Introduction
*IsoBayes* is a Bayesian method to perform inference on single protein isoforms.
Our approach infers the presence/absence of protein isoforms, and also estimates their abundance;
additionally, it provides a measure of the uncertainty of these estimates, via:
i) the posterior probability that a protein isoform is present in the sample;
ii) a posterior credible interval of its abundance.
*IsoBayes* inputs liquid cromatography mass spectrometry (MS) data,
and can work with both PSM counts, and intensities.
When available, trascript isoform abundances (i.e., TPMs) are also incorporated:
TPMs are used to formulate an informative prior for the respective protein isoform relative abundance.
We further identify isoforms where the relative abundance of proteins and transcripts significantly differ.
We use a two-layer latent variable approach to model two sources of uncertainty typical of MS data:
i) peptides may be erroneously detected (even when absent);
ii) many peptides are compatible with multiple protein isoforms.
In the first layer, we sample the presence/absence of each peptide based on its estimated probability 
of being mistakenly detected, also known as PEP (i.e., posterior error probability).
In the second layer, for peptides that were estimated as being present, 
we allocate their abundance across the protein isoforms they map to.
These two steps allow us to recover the presence and abundance of each protein isoform.

## Bioconductor installation
*IsoBayes* is available on [Bioconductor](https://bioconductor.org/packages/IsoBayes) and can be installed with the command:
```{r Bioconductor_installation, eval=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
  install.packages("BiocManager")
BiocManager::install("IsoBayes")
```

To access the R code used in the vignettes, type:
```{r vignettes, eval=FALSE} 
browseVignettes("IsoBayes")
```

# Questions, issues and citation
Questions relative to *IsoBayes* should be reported as a new issue at *[BugReports](https://github.com/SimoneTiberi/IsoBayes/issues)*.

To cite *IsoBayes*, type:
```{r citation} 
citation("IsoBayes")
```

# Load the package
Load *IsoBayes*:
```{r load, message=FALSE}
library(IsoBayes)
```

# Key options

## Input data
*IsoBayes* works with the output of both *MetaMorpheus* (MM) [@MetaMorpheus], or *Percolator* [@Percolator] (via the *OpenMS* toolkit [@OpenMS]).
Other tools can be used as well to generate out input data, as long as the input files follow the structure mentioned in the [Input user-provided data](#input-user-provided-data) Section.
In our benchmarks, we tested our model using both *MetaMorpheus* and *Percolator* data, and obtained slightly better results, and a shorter runtime with *MetaMorpheus*.

In this tutorial, we use a small dataset created by *MetaMorpheus*.
The code to generate the data can be found [here](https://github.com/SimoneTiberi/IsoBayes/blob/main/inst/script/extdata.Rmd).

## PSM counts vs. intensities
We can run *IsoBayes* on either PSM counts or intensities.
In our benchmarks, we found that, although results are consistent between the two types of input data, intensities led to a marginal improvement in performance.

## TPMs (optional)
If available, transcriptomics data (from short of long reads), can also be integrated in the model in the form of TPMs; this will enhance protein isoform inference.
To correctly match proteomics and transcriptomics data, transcript isoform and protein isoform names must be consistent.
Here, we also specify the path to TPMs (optional); which must be stored in a `.tsv` file with 1 row per isoform, and 2 columns:

* 'isoname': a character string indicating the isoform name;
* 'tpm': a numeric variable indicating the TPM counts.
Column names must be 'isoname' and 'tpm'.

## PEP vs. FDR cutoff
Additionally, we suggest to provide the probability that peptides are erroneously detected (usually called PEP): *IsoBayes* will use the PEP to sample the presence/absence of each peptide; this will propagate the peptide identification uncertainty throughout the inference process.
Alternatively, unreliable peptides can be discarded by specifying a peptide False Discovery Rate (FDR) threshold (also known as qvalue).
In our benchmarks, we found that using `PEP` provides a (minor) increase in performace, at the cost of a (marginally) higher runtime.
By default, \code{load_data} uses the PEP (`PEP = TRUE`); if we want to disable the PEP integration, and filter based on the FDR, we can set `PEP = FALSE` and specify an FDR cutoff, e.g., as `FDR_thd = 0.01`.

# Load the data
First, we load and process the input data using the \code{load_data} function.

## Input *MetaMorpheus* data
We set the directory of the data (internal in the package):
```{r specify extdata path}
data_dir = system.file("extdata", package = "IsoBayes")
```

We set the path to the AllPeptides.psmtsv file returned by *MetaMorpheus*:
```{r set path MM psm}
path_to_peptides_psm = paste0(data_dir, "/AllPeptides.psmtsv")
```

We set the path (optional) to the TPMs:
```{r set path mRNA}
tpm_path = paste0(data_dir, "/jurkat_isoform_kallisto.tsv")
```

### Peptide-Spectrum Match (PSM) counts
The AllPeptides.psmtsv file contains all the information required to run *IsoBayes* with PSM counts;
we load the file via \code{load_data}:
```{r load psm}
PSM_data_loaded = load_data(path_to_peptides_psm = path_to_peptides_psm,
                        path_to_tpm = tpm_path)
```

### Intensities
If we want to run the algorithm with peptide intensities, in addition to AllPeptides.psmtsv, we also need to load a second file generated by MM: AllQuantifiedPeptides.tsv.
Then, we have to set `abundance_type` argument to "intensities"; otherwise, the \code{load_data} function will ignore the AllQuantifiedPeptides.tsv file.
```{r set path MM intensities}
path_to_peptides_intensities = paste0(data_dir, "/AllQuantifiedPeptides.tsv")
data_loaded = load_data(path_to_peptides_psm = path_to_peptides_psm,
                        path_to_peptides_intensities = path_to_peptides_intensities,
                        abundance_type = "intensities",
                        path_to_tpm = tpm_path)
```

## Input *Percolator* data
*IsoBayes* is also compatible with the PSM file from *Percolator* (from the *OpenMS* toolkit).
The [README](https://github.com/SimoneTiberi/IsoBayes/tree/main#readme) shows how to use *OpenMS* tools to generate an idXML file containing PSM data.
Once the file has been created, we can pass its path to \code{load_data} and set `input_type` as "openMS".
```{r set path OpenMS, eval = FALSE}
Percolator_data_loaded = load_data(path_to_peptides_psm = "/path/to/file.idXML",
                        input_type = "openMS",
                        path_to_tpm = tpm_path)
```
Please note that when using data generated by *Percolator*, the algorithm can only process PSM counts and not intensities.

## Input user-provided data
We can also input MS data obtained from any bioinformatics tool.
To this aim, the data must be organized in a `.tsv` file where each row corresponds to a peptide with up to four fields:

* 'Y': a numeric variable indicating the peptide abundance (PSM counts or intensities, as defined by the user);
* 'EC': Equivalent Classes, a character string indicating the isoform(s) name the peptide maps to. If the peptide maps to multiple protein isoforms, the names must be separated with "|" , i.e. "name_isoform_1|name_isoform_2";
* 'FDR': (optional) a numeric variable indicating the FDR of each peptide;
* 'PEP': (optional) a numeric variable indicating the probability that a peptide is erroneously detected;
* 'sequence': (required when using PEP) a character string indicating the peptide name/id/amino acids sequence.

Note that, when using user-provided data, arguments `abundance_type` and `path_to_peptides_intensities` are not considered, because users are free to set column `Y` to PSM counts or intensities.

To load the tsv file with user data, we just need to pass the file path and set `input_type` as "other" .
```{r set path user, eval = FALSE}
Other_data_loaded = load_data(path_to_peptides_psm = "/path/to/user_data.tsv",
                        input_type = "other",
                        path_to_tpm = tpm_path)
```

# Inference
Once we have loaded the data, we can run the algorithm using the `inference` function.
```{r inference, eval = FALSE}
set.seed(169612)
results = inference(data_loaded)
```

By default, *IsoBayes* uses one single core.
For large datasets, to speed up the execution time, the number of cores can be increased via the `n_cores` argument.

## Gene Normalization (optional)
In order to analyse alternative splicing within a specific gene, we may want to normalize the estimated relative abundances for each set of protein isoforms that maps to a gene. To this aim, we need to input a csv file containing two columns denoting the isoform name and the gene name.
```{r default inference}
path_to_map_iso_gene = paste0(data_dir, "/map_iso_gene.csv")
set.seed(169612)
results_normalized = inference(data_loaded, map_iso_gene = path_to_map_iso_gene)
```

Specifying the `map_iso_gene` argument, also enables users to plot results via `plot_relative_abundances` function.

# Visualizing results
Results are stored as a `list` with two `data.frame` objects: 'isoform_results' and 'normalized_isoform_results' (NULL if 'map_iso_gene' was not provided).
```{r get res}
names(results_normalized)
```

Inside the 'isoform_results' `data.frame`, each raw corresponds to a protein isoform. The columns report the following results:

* 'Prob_present': protein isoform probability of being present in the biological sample;
* 'Abundance': estimation of the protein isoform absolute abundance;
* 'CI_LB': lower bound of the highest posterior density interval estimation of the protein isoform absolute abundance;
* 'CI_UB': upper bound of the highest posterior density interval estimation of the protein isoform absolute abundance;
* 'Pi': estimation of the protein isoform relative abundance;
* 'Pi_CI_LB': lower bound of the highest posterior density interval estimation of the protein isoform relative abundance;
* 'Pi_CI_UB': upper bound of the highest posterior density interval estimation of the protein isoform relative abundance;
* 'TPM': transcript isoform abundance;
* 'Log2_FC' : logarithm to the base 2 of the ratio between 'Pi' and 'TPM' normalized, indicating how much the isoform abundance changed after transcription;
* 'Prob_prot_inc' : probability that the abundance of the protein isoform is greater than the abundance of the transcript isoform.

```{r get iso res}
head(results_normalized$isoform_results)
```

In 'normalized_isoform_results' `data.frame`, we report protein isoform relative abundances, normalized within each gene (i.e., adding to 1 within a gene).
```{r get iso res norm}
head(results_normalized$normalized_isoform_results)
```

Finally, *IsoBayes* provides the `plot_relative_abundances` function to visualize differential relative abundances between transcriptomics and proteomics isoforms within a specific gene:
```{r plotting results}
plot_relative_abundances(results_normalized, gene_id = "TUBB")
```

By default `plot_relative_abundances` normalizes the relative abundances.
To disable the normalization, we just need to set as `FALSE` the
`normalize_gene` argument.
```{r plotting results no normalization}
plot_relative_abundances(results_normalized, gene_id = "TUBB", normalize_gene = FALSE)
```

Note that `plot_relative_abundances` can be used only when, in the `map_iso_gene` argument of the `inference` function, we provide a csv file that maps the protein isoforms to the corresponding gene (`path_to_map_iso_gene` in this case).

# Session info
```{r sessionInfo}
sessionInfo()
```

# References